File: PreUtil/src/output/java/hard/BestTimeToBuyAndSellStockIii.java
/**https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/ */
//You are given an array prices where prices[i] is the price of a given stock on the ith day.
//Find the maximum profit you can achieve. You may complete at most two transactions.
//Note: You may not engage in multiple transactions simultaneously (i.e., you must sell the stock before you buy again).


class BestTimeToBuyAndSellStockIii {
    public int maxProfit(int[] prices) {
        int n = prices.length;
        int[] left = new int[n];
        int[] right = new int[n];
        int min = prices[0];
        for (int i = 1; i < n; i++) {
            min = Math.min(min, prices[i]);
            left[i] = Math.max(left[i - 1], prices[i] - min);
        }
        int max = prices[n - 1];
        for (int i = n - 2; i >= 0; i--) {
            max = Math.max(max, prices[i]);
            right[i] = Math.max(right[i + 1], max - prices[i]);
        }
        int ans = 0;
        for (int i = 0; i < n; i++) {
            ans = Math.max(ans, left[i] + right[i]);
        }
        return ans;
        
    }
}
     
    

File: Tool/PreUtil/src/output/javascript/hard/BestTimeToBuyAndSellStockIii.js
/**https://leetcode.com/problems/best-time-to-buy-and-sell-stock-iii/ */
//You are given an array prices where prices[i] is the price of a given stock on the ith day.
//Find the maximum profit you can achieve. You may complete at most two transactions.
//Note: You may not engage in multiple transactions simultaneously (i.e., you must sell the stock before you buy again).


/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    let n = prices.length;
    let dp = [];
    for(let i = 0; i <= n; i++){
        dp.push([]);
        for(let j = 0; j <= 2; j++){
            dp[i].push([]);
            for(let k = 0; k <= 1; k++){
                dp[i][j].push(0);
            }
        }
    }
    for(let i = 0; i <= n; i++){
        for(let j = 0; j <= 2; j++){
            dp[i][j][0] = 0;
            dp[i][j][1] = -Infinity;
        }
    }
    for(let i = 1; i <= n; i++){
        for(let j = 1; j <= 2; j++){
            dp[i][j][0] = Math.max(dp[i - 1][j][0], dp[i - 1][j][1] + prices[i - 1]);
            dp[i][j][1] = Math.max(dp[i - 1][j][1], dp[i - 1][j - 1][0] - prices[i - 1]);
        }
    }
    return dp[n][2][0];
    
};

    